chalice/local.py

```diff
@@ -109,98 +181,246 @@ def create_lambda_event(self, method, path, headers, body=None):
+        # by your Lambda function.
+        self.log_group_name = ''
+
+        # The name of the CloudWatch log stream where you can find logs
+        # written by your Lambda function. The log stream may or may not
+        # change for each invocation of the Lambda function.
+        #
+        # The value is null if your Lambda function is unable to create a log
+        # stream, which can happen if the execution role that grants necessary
+        # permissions to the Lambda function does not include permissions for
+        # the CloudWatch Logs actions.
+        self.log_stream_name = ''
+
+        # The last two attributes have the following comment in the
+        # documentation:
+        # Information about the client application and device when invoked
+        # through the AWS Mobile SDK, it can be null.
+        # Chalice local doens't need to set these since they are specifically
+        # for the mobile SDK.
+        self.identity = None
+        self.client_context = None
+
+    def _current_time_millis(self):
+        # type: () -> float
+        return self._time_source.time() * 1000
+
+    def get_remaining_time_in_millis(self):
+        # type: () -> float
+        runtime = self._current_time_millis() - self._start_time
+        return self._max_runtime - runtime
+
+
+class LocalGatewayAuthorizer(object):
+    """A class for running user defined authorizers in local mode."""
+    def __init__(self, app_object):
+        # type: (Chalice) -> None
+        self._app_object = app_object
+        self._arn_builder = LocalArnBuilder()
+
+    def authorize(self, lambda_event, lambda_context):
+        # type: (EventType, LambdaContext) -> Tuple[EventType, LambdaContext]
+        method = lambda_event['requestContext']['httpMethod']
+        path = lambda_event['requestContext']['resourcePath']
+        arn = self._arn_builder.build_arn(method, path)
+        auth_event = self._prepare_authorizer_event(arn, lambda_event)
+        route_entry = self._route_for_event(lambda_event)
+        if not route_entry:
+            return lambda_event, lambda_context
+        authorizer = route_entry.authorizer
+        if not authorizer:
+            return lambda_event, lambda_context
+        if not isinstance(authorizer, ChaliceAuthorizer):
+            # Currently the only supported local authorizer is the
+            # BuiltinAuthConfig type. Anything else we will err on the side of
+            # allowing local testing by simply admiting the request. Otherwise
+            # there is no way for users to test their code in local mode.
+            warnings.warn(
+                '%s is not a supported in local mode. All requests made '
+                'against a route will be authorized to allow local testing.'
+                % authorizer.__class__.__name__
+            )
+            return lambda_event, lambda_context
+        auth_result = authorizer(auth_event, lambda_context)
```

Comment:
I don't think this is quite right.  If you don't have the appropriate auth header (`Authorization`) in the request, you get a 401 and the authorizer isn't called at all.